lib/IRGen/GenBuiltin.cpp

```diff
@@ -1543,3 +1545,683 @@ void irgen::emitBuiltinCall(IRGenFunction &IGF, const BuiltinInfo &Builtin,
+    emitDestroyCall(IGF, T, stackAddress.getAddress());
+    deallocateStack(IGF, stackAddress, T);
+  }
+
+  TypeLayoutEntry *
+  buildTypeLayoutEntry(IRGenModule &IGM,
+                       SILType T,
+                       bool useStructLayouts) const override {
+    return IGM.typeLayoutCache.getOrCreateResilientEntry(T);
+  }
+
+  llvm::Value *getEnumTagSinglePayload(IRGenFunction &IGF,
+                                       llvm::Value *numEmptyCases,
+                                       Address arrayAddr,
+                                       SILType arrayType,
+                                       bool isOutlined) const override {
+    // take extra inhabitants from the first element
+    auto firstElementAddr
+      = IGF.Builder.CreateElementBitCast(arrayAddr, Element.getStorageType());
+    return Element.getEnumTagSinglePayload(IGF,
+                                           numEmptyCases,
+                                           firstElementAddr,
+                                           getElementSILType(IGF.IGM, arrayType),
+                                           isOutlined);
+  }
+
+  void storeEnumTagSinglePayload(IRGenFunction &IGF,
+                                 llvm::Value *index,
+                                 llvm::Value *numEmptyCases,
+                                 Address arrayAddr,
+                                 SILType arrayType,
+                                 bool isOutlined) const override {
+    // take extra inhabitants from the first element
+    auto firstElementAddr
+      = IGF.Builder.CreateElementBitCast(arrayAddr, Element.getStorageType());
+    return Element.storeEnumTagSinglePayload(IGF,
+                                          index,
+                                          numEmptyCases,
+                                          firstElementAddr,
+                                          getElementSILType(IGF.IGM, arrayType),
+                                          isOutlined);
+  }
+};
+
+const TypeInfo *
+TypeConverter::convertBuiltinFixedArrayType(BuiltinFixedArrayType *T) {
+  // Most of our layout properties come from the element type.
+  auto &elementTI = IGM.getTypeInfoForUnlowered(AbstractionPattern::getOpaque(),
+                                                T->getElementType());
+  // ...unless the array size is not fixed, then the array layout is never
+  // fixed.
+  auto fixedSize = T->getFixedSize();
+  
+  // Statically zero or negative-sized array types are empty.
+  if (fixedSize == 0 || T->isFixedNegativeSize()) {
+    return &getEmptyTypeInfo();
+  }
+                                  
+  if (!fixedSize.has_value() || !elementTI.isFixedSize()) {
+    return new NonFixedArrayTypeInfo(IGM.OpaqueTy, elementTI);
+  }
+  
+  if (auto *loadableTI = dyn_cast<LoadableTypeInfo>(&elementTI)) {
+    return new LoadableArrayTypeInfo(fixedSize.value(), *loadableTI);
+  }
+  
+  return new FixedArrayTypeInfo(fixedSize.value(),
```

Comment:
Does this work if our size is fixed, but the element is still generic?

Reply:
Yeah, it should, since we're only fixed size iff the element size is fixed size, and we check `!elementTI.isFixedSize()` above.

Reply:
Oh I didn't see that condition 👍